File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Loops.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace Monitor_Loops
{
    class Program
    {
        private object obj = new object();
        private object other = new object();
        private bool condition;

        public void Method1()
        {
            Monitor.Enter(obj);     // Noncompliant tricky FP, as the execution should always reach number 9, but we don't track that
            if (condition)
            {
                Monitor.Exit(obj);  // To release it on at least one path to activate the rule
            }
            for (int i = 0; i < 10; i++)
            {
                if (i == 9)
                {
                    Monitor.Exit(obj);
                }
            }
            Monitor.Enter(other);   // Noncompliant
            if (condition) {
                Monitor.Exit(other);
            }
        }

        public void Method2()
        {
            Monitor.Enter(obj); // Compliant, because the lock is not released on any path
            for (int i = 0; i < 10; i++)
            {
                break;
                if (i == 9)
                {
                    Monitor.Exit(obj);
                }
            }
            Monitor.Enter(other);   // Noncompliant, to make sure we explore paths after the loop
            if (condition)
            {
                Monitor.Exit(other);
            }
        }

        public void Method3()
        {
            Monitor.Enter(obj); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);  // To release it on at least one path to activate the rule
            }
            for (int i = 0; i < 10; i++)
            {
                if (i == 5)
                {
                    break;
                }

                if (i == 9)
                {
                    Monitor.Exit(obj);
                }
            }
            Monitor.Enter(other);   // Noncompliant
            if (condition)
            {
                Monitor.Exit(other);
            }
        }

        public void Method4()
        {
            Monitor.Enter(obj); // Noncompliant tricky FP, as the execution should always reach number 9, but we don't track that
            if (condition)
            {
                Monitor.Exit(obj);  // To release it on at least one path to activate the rule
            }
            for (int i = 0; i < 10; i++)
            {
                if (i == 10)
                {
                    break;
                }

                if (i == 9)
                {
                    Monitor.Exit(obj);
                }
            }
            Monitor.Enter(other);   // Noncompliant
            if (condition)
            {
                Monitor.Exit(other);
            }
        }

        public void Method5()
        {
            Monitor.Enter(obj); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);  // To release it on at least one path to activate the rule
            }
            for (int i = 0; i < 10; i++)
            {
                if (i == 9)
                {
                    continue;
                }

                if (i == 9)
                {
                    Monitor.Exit(obj);
                }
            }
            Monitor.Enter(other);   // Noncompliant
            if (condition)
            {
                Monitor.Exit(other);
            }
        }

        public void Method6(bool condition, byte[] array)
        {
            Monitor.Enter(obj); // Noncompliant
            foreach (var item in array)
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
            }
        }

        public void Method7(bool condition, byte[] array)
        {
            Monitor.Enter(obj); // Noncompliant, array can be empty
            foreach (var item in array)
            {
                Monitor.Exit(obj);
            }
        }

        public void Method8(bool condition, List<byte> array)
        {
            Monitor.Enter(obj); // Noncompliant
            while (array.Count < 42)
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
                array.RemoveAt(0);
            }
        }

        public void Method9(bool condition, List<byte> array)
        {
            Monitor.Enter(obj); // Noncompliant, count can be bigger than 42
            while (array.Count < 42)
            {
                Monitor.Exit(obj);
                array.RemoveAt(0);
            }
        }

        public void Method10(bool condition, List<byte> array)
        {
            Monitor.Enter(obj); // Noncompliant
            do
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
                array.RemoveAt(0);
            }
            while (array.Count < 42);
        }

        public void Method11(bool condition, List<byte> array)
        {
            Monitor.Enter(obj); // Compliant
            do
            {
                Monitor.Exit(obj);
                array.RemoveAt(0);
            }
            while (array.Count < 42);
        }

        public void For_CanExit_PreIncrement()
        {
            for (var i = 0; i < 10; ++i)
            {
                Console.WriteLine();
            }

            Monitor.Enter(obj); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }

        public void For_CanExit_PostIncrement()
        {
            for (var i = 0; i < 10; i++)
            {
                Console.WriteLine();
            }

            Monitor.Enter(obj); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }

        public void For_CanExit_PreDecrement()
        {
            for (var i = 10; i > 0; --i)
            {
                Console.WriteLine();
            }

            Monitor.Enter(obj); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }

        public void For_CanExit_PostDecrement()
        {
            for (var i = 10; i > 0; i--)
            {
                Console.WriteLine();
            }

            Monitor.Enter(obj); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Loops.vb
Imports System.Threading

Namespace Monitor_Loops

    Class Program

        Private Condition As Boolean
        Private Obj As New Object()
        Private Other As New Object()

        Public Sub Method1()
            Monitor.Enter(Obj)      ' Noncompliant tricky FP, as the execution should always reach number 9, but we don't track that
            For i As Integer = 0 To 9
                If i = 9 Then Monitor.Exit(Obj)
            Next
            Monitor.Enter(Other)    ' Noncompliant
            If Condition Then Monitor.Exit(Other)
        End Sub

        Public Sub Method2()
            Monitor.Enter(Obj) ' FN
            For i As Integer = 0 To 9
                Exit For
                If i = 9 Then Monitor.Exit(Obj)
            Next
        End Sub

        Public Sub Method3()
            Monitor.Enter(Obj)      ' Noncompliant
            For i As Integer = 0 To 9
                If i = 5 Then Exit For
                If i = 9 Then Monitor.Exit(Obj)
            Next
            Monitor.Enter(Other)    ' Noncompliant
            If Condition Then Monitor.Exit(Other)
        End Sub

        Public Sub Method4()
            Monitor.Enter(Obj)      ' Noncompliant tricky FP, as the execution should always reach number 9, but we don't track that
            For i As Integer = 0 To 9
                If i = 10 Then Exit For
                If i = 9 Then Monitor.Exit(Obj)
            Next
            Monitor.Enter(Other)    ' Noncompliant
            If Condition Then Monitor.Exit(Other)
        End Sub

        Public Sub Method5()
            Monitor.Enter(Obj)      ' Compliant, Exit is not reached on any path. Should be covered by S2583.
            For i As Integer = 0 To 9
                If i = 9 Then Continue For
                If i = 9 Then Monitor.Exit(Obj)
            Next
            Monitor.Enter(Other)    ' Noncompliant
            If Condition Then Monitor.Exit(Other)
        End Sub

        Public Sub Method6(Array() As Byte)
            Monitor.Enter(Obj)      ' Noncompliant
            For Each Item As Byte In Array
                If Condition Then Monitor.Exit(Obj)
            Next
        End Sub

        Public Sub Method7(Array() As Byte)
            Monitor.Enter(Obj)      ' Noncompliant, array can be empty
            For Each Item In Array
                Monitor.Exit(Obj)
            Next
        End Sub

        Public Sub Method8(Array As List(Of Byte))
            Monitor.Enter(Obj)      ' Noncompliant
            While Array.Count < 42
                If Condition Then Monitor.Exit(Obj)
                Array.RemoveAt(0)
            End While
        End Sub

        Public Sub Method9(Array As List(Of Byte))
            Monitor.Enter(Obj)      ' Noncompliant, count can be bigger than 42
            While Array.Count < 42
                Monitor.Exit(Obj)
                Array.RemoveAt(0)
            End While
        End Sub

        Public Sub Method10(Array As List(Of Byte))
            Monitor.Enter(Obj)      ' Noncompliant
            Do
                If Condition Then Monitor.Exit(Obj)
                Array.RemoveAt(0)
            Loop While Array.count < 42
        End Sub

        Public Sub Method11(Array As list(Of Byte))
            Monitor.Enter(Obj)      ' Compliant
            Do
                Monitor.Exit(Obj)
                Array.RemoveAt(0)
            Loop While Array.Count < 42
        End Sub

    End Class

End Namespace
